CardinalityEstimation/CardinalityEstimator.cs

```diff
@@ -118,6 +118,38 @@ public CardinalityEstimator(int b = 14)
             }
         }
 
+        internal CardinalityEstimator(CardinalityEstimatorData data)
+        {
+            this.bitsPerIndex = data.bitsPerIndex;
+            this.bitsForHll = 64 - this.bitsPerIndex;
+            this.m = (int)Math.Pow(2, this.bitsPerIndex);
+            this.alphaM = GetAlphaM(this.m);
+            this.subAlgorithmSelectionThreshold = GetSubAlgorithmSelectionThreshold(this.bitsPerIndex);
+
+            // Init the sparse representation
+            this.isSparse = data.isSparse;
+            this.lookupSparse = data.lookupSparse != null ? new Dictionary<ushort, byte>(data.lookupSparse) : null;
+            this.lookupDense = data.lookupDense;
+
+            // Each element in the sparse representation takes 15 bytes, and there is some constant overhead
+            this.sparseMaxElements = Math.Max(0, this.m / 15 - 10);
+            // If necessary, switch to the dense representation
```

Comment:
This code (136-150) seems to depend on the serialization implementation rather than just on CardinalityEstimatorData (it accounts for not serializing lookupSparse when directCount != null).  We should either:
1. Move it to the deserializer, and run it before using this constructor (which means making AddElementHash available there, which I prefer to avoid), or
2. Move the ser/deserialization logic into CardinalityEstimator by implementing ISerializable